src/module/types.rs

```diff
@@ -6,22 +6,66 @@ use crate::error::Result;
 use crate::module::Module;
 use crate::parse::IndicesToIds;
 use crate::ty::{Type, TypeId, ValType};
+use std::sync::{Arc, Mutex};
 
 /// The set of de-duplicated types within a module.
+///
 #[derive(Debug, Default)]
 pub struct ModuleTypes {
-    arena: ArenaSet<Type>,
+    arena: Mutex<ArenaSet<Type>>,
 }
 
+/// Low-level operations for working with `ModuleTypes`.
+///
+/// # Internal Mutex
+///
+/// The types within a `ModuleTypes` are protected by a mutex, and the methods
+/// in this `impl` block all lock the mutex for the duration of the closure's
+/// execution. **This means that re-entrancy within these closure-taking methods
+/// will dead lock, and it is your responsibility to avoid it.**
+///
+/// Generally, prefer using the higher-level methods on `ModuleTypes` that don't
+/// take closures and hide the details of the mutex from you.
 impl ModuleTypes {
-    /// Get a type associated with an ID
-    pub fn get(&self, id: TypeId) -> &Type {
-        &self.arena[id]
+    /// Get a type associated with an ID.
+    pub fn with<T>(&self, id: TypeId, f: impl FnOnce(&Type) -> T) -> T {
+        let arena = self.arena.lock().unwrap();
+        f(&arena[id])
     }
 
-    /// Get a type associated with an ID
-    pub fn get_mut(&mut self, id: TypeId) -> &mut Type {
-        &mut self.arena[id]
+    /// Get a type associated with an ID mutably.
+    pub fn with_mut<T>(&self, id: TypeId, f: impl FnOnce(&mut Type) -> T) -> T {
+        let mut arena = self.arena.lock().unwrap();
+        f(&mut arena[id])
+    }
+
+    /// Iterate over each of this module's types.
+    pub fn each(&self, mut f: impl FnMut(&Type)) {
```

Comment:
This is sort of a bummer to have, but I wonder if we could have a `lock()` method or something like that which then provides iteration on the returned type?